src/gpu/nvidia/sycl_cuda_memory_storage_helper.hpp

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
+* Copyright 2022 Codeplay Software Limited
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_NVIDIA_SYCL_CUDA_MEMORY_STORAGE_HELPER_HPP
+#define GPU_NVIDIA_SYCL_CUDA_MEMORY_STORAGE_HELPER_HPP
+
+#include <optional>
+#include "gpu/nvidia/sycl_cuda_scoped_context.hpp"
+#include "sycl/sycl_memory_storage.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace nvidia {
+
+#define CTX_IN_SYCL_MEMORY(arg) \
+    sycl_memory_arg<::sycl::access::mode::read>( \
+            static_cast<sycl::sycl_memory_storage_base_t *>( \
+                    &CTX_IN_STORAGE(arg)), \
+            cgh)
+
+#define CTX_OUT_SYCL_MEMORY(arg) \
+    sycl_memory_arg<::sycl::access::mode::write>( \
+            static_cast<sycl::sycl_memory_storage_base_t *>( \
+                    &CTX_OUT_STORAGE(arg)), \
+            cgh)
+
+template <::sycl::access_mode mode>
+class sycl_memory_arg {
+public:
+    sycl_memory_arg() = default;
+    sycl_memory_arg(
+            sycl::sycl_memory_storage_base_t *mem, ::sycl::handler &cgh) {
+        switch (mem->memory_kind()) {
+            case sycl::memory_kind::buffer:
+                acc_.emplace(
+                        utils::downcast<sycl::sycl_buffer_memory_storage_t *>(
+                                mem)
+                                ->buffer(),
+                        cgh);
+                break;
+            case sycl::memory_kind::usm:
+                usm_mem_ = utils::downcast<
+                        const sycl::sycl_usm_memory_storage_t *>(mem);
+                break;
+            default: assert(!"unexpected memory kind");
+        }
+    }
+
+    template <typename T = void>
+    T *get_native_pointer(const compat::interop_handle &ih,
+            const cuda_sycl_scoped_context_handler_t &sc) const {
+        void *raw_ptr;
+        if (usm_mem_) {
+            raw_ptr = usm_mem_->usm_ptr();
+        } else {
+            raw_ptr = sc.memory<void *>(ih, acc_.value());
+        }
+        return reinterpret_cast<T *>(raw_ptr);
+    }
+
+    template <typename T = void>
+    T *get_native_pointer(const compat::interop_handle &ih,
```

Comment:
What is the reason for explicitly applying the offset? We can have offset as a data member that would be initialized in the constructor by the value returned by `base_offset()` for buffers and by 0 for USM.
Then we can have just one `get_native_pointer(...)` function that would implicitly apply the offset.

Reply:
We need to explicitly pass it it either into the constructor or into this function. This option makes the implementation a bit simpler by not storing a copy of the offset internally. 

Also, many arguments, including all in `cudnn_binary` do not use explicit offset, so this lets us handle that use case by not dealing with offset at all.

Reply:
But the constructor already takes `sycl_memory_storage_base_t` that contains the offset if it points to `sycl_buffer_memory_storage_t`. We just need to query it using `base_offset()` and then initialize a new `offset_` data member with the queried value. For USM it will always be 0 (there is no `base_offset()` for `sycl_usm_memory_storage_t`), for buffers it will always be 0 except the cases when a memory storage is actually a sub-storage (e.g. for scratchpad).

> Also, many arguments, including all in cudnn_binary do not use explicit offset, so this lets us handle that use case by not dealing with offset at all.

In such cases the offset will always be 0 therefore the returned native pointer won't be shifted. 

My main concern here is that applying the offset is basically a workaround for the bug that will be fixes at some point. This is why I think it would be better to handle it in one place inside `sycl_memory_arg_t` rather than in each primitive that has to deal with offsets.

Reply:
done

Reply:
Looks good to me, thanks.